tokio/src/runtime/task/list.rs

```diff
@@ -111,25 +110,30 @@ impl<S: 'static> OwnedTasks<S> {
             // to the field.
             task.header().set_owner_id(self.id);
         }
-
-        let mut lock = self.inner.lock();
-        if lock.closed {
-            drop(lock);
-            drop(notified);
+        // check close flag
+        if self.closed.load(Ordering::Acquire) {
             task.shutdown();
-            None
-        } else {
-            lock.list.push_front(task);
-            Some(notified)
+            return None;
         }
+        self.list.push(task);
+
+        // double check close flag for ensuring all tasks will shutdown after OwnedTasks has been closed,
+        // it should be completed quickly.
+        if self.closed.load(Ordering::Acquire) {
+            for i in 0..self.get_shard_size() {
+                if let Some(task) = self.list.pop_back(i) {
+                    task.shutdown();
+                }
+            }
+        }
```

Comment:
You can avoid this by defining a method like this:
```
impl<L: ShardedListItem> ShardedList<L, L::Target> {
    pub fn lock_shard(&self, val: &L::Handle) -> ShardGuard<'_, L, L::Target> {
        let id = unsafe { L::get_shared_id(L::as_raw(&val)) };
        ShardGuard {
            lock: self.shard_inner(id),
            count: &self.count,
            id,
        }
    }
}

struct ShardGuard<'a, L, T> {
    lock: MutexGuard<'a, LinkedList<L, T>>
    count: &'a AtomicUsize,
    id: usize,
}

impl<L: ShardedListItem> ShardGuard<L, L::Target> {
    /// Push a value to this shard.
    pub fn push_shard(self, val: L::Handle) {
        let id = unsafe { L::get_shared_id(L::as_raw(&val)) };
        assert_eq!(id, self.id);
        self.lock.push_front(val);
        self.count.fetch_add(1, Ordering::Relaxed);
    }
}
```
This will let you check `self.closed` while the mutex is locked.